File: app/src/main/java/com/github/deskid/focusreader/widget/WebImageView.java
package com.github.deskid.focusreader.widget;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.support.annotation.Nullable;
import android.support.v7.widget.AppCompatImageView;
import android.text.TextUtils;
import android.util.AttributeSet;

import com.squareup.picasso.Picasso;

public class WebImageView extends AppCompatImageView {

    private String mImageUrl;
    private int mResource;


    public WebImageView(Context context) {
        this(context, null);
    }

    public WebImageView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    public void setImageResource(int imageResource) {
        super.setImageResource(imageResource);
        mResource = imageResource;
        mImageUrl = null;
    }

    @Override
    public void setImageDrawable(@Nullable Drawable drawable) {
        super.setImageDrawable(drawable);
    }

    public void setImageUrl(String url) {
        if (TextUtils.isEmpty(url)) {
            mImageUrl = null;
            Picasso.with(getContext()).cancelRequest(this);
            setImageBitmap(null);
        } else {
            mImageUrl = url;
            if (isAttachedToWindow()) {
                Picasso.with(getContext()).load(Uri.parse(url)).into(this);
            }
        }
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        if (!TextUtils.isEmpty(mImageUrl)) {
            setImageUrl(mImageUrl);
        } else if (mResource != -1) {
            setImageResource(mResource);
        }

    }

    @Override
    protected void onDetachedFromWindow() {
        Picasso.with(getContext()).cancelRequest(this);
        super.onDetachedFromWindow();
    }
}

File: app/src/main/java/com/github/deskid/focusreader/widget/WebImageView.kt
package com.github.deskid.focusreader.widget

import android.content.Context
import android.graphics.drawable.Drawable
import android.support.v7.graphics.Palette
import android.support.v7.widget.AppCompatImageView
import android.util.AttributeSet
import android.util.Log
import com.bumptech.glide.Glide
import com.bumptech.glide.Priority
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.Target
import com.github.florent37.glidepalette.BitmapPalette
import com.github.florent37.glidepalette.GlidePalette

class WebImageView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) : AppCompatImageView(context, attrs) {

    private var mImageUrl: String? = null
    private var mResource: Int = 0

    override fun setImageResource(imageResource: Int) {
        super.setImageResource(imageResource)
        mResource = imageResource
        mImageUrl = null
    }

    override fun setImageDrawable(drawable: Drawable?) {
        super.setImageDrawable(drawable)
    }

    fun setImageUrl(url: String?, onPaletteLoaded: ((palette: Palette?) -> Unit)? = null, onImageLoaded: (() -> Unit)? = null) {
        Log.d("TAG", "onPaletteLoaded=" + onPaletteLoaded + "onImageLoaded=" + onImageLoaded)
        if (url.isNullOrEmpty()) {
            mImageUrl = null
            setImageBitmap(null)
        } else {
            mImageUrl = url

            val option = RequestOptions()
                    .dontAnimate()
                    .priority(Priority.HIGH)

            val requestListener = object : RequestListener<Drawable> {
                override fun onLoadFailed(e: GlideException?, model: Any?, target: Target<Drawable>?, isFirstResource: Boolean): Boolean {
                    onImageLoaded?.invoke()
                    return false
                }

                override fun onResourceReady(resource: Drawable?, model: Any?, target: Target<Drawable>?, dataSource: DataSource?, isFirstResource: Boolean): Boolean {
                    onImageLoaded?.invoke()
                    return false
                }
            }

            if (onPaletteLoaded == null) {
                Glide.with(this)
                        .load(url)
                        .apply(option)
                        .listener(requestListener)
                        .into(this)
            } else {
                Glide.with(this)
                        .load(url)
                        .apply(option)
                        .listener(
                                GlidePalette
                                        .with(url)
                                        .use(BitmapPalette.Profile.MUTED_DARK)
                                        .setGlideListener(requestListener)
                                        .intoCallBack(onPaletteLoaded))
                        .into(this)
            }
        }
    }
}
